microprofile/jwt-auth/jwt-auth/src/main/java/io/helidon/microprofile/jwt/auth/JwtAuthAnnotationAnalyzer.java

```diff
@@ -50,25 +56,89 @@ public void init(Config config) {
                         });
                     });
                 });
+        secureByDefault = config.get("jwt.secure-by-default")
+                .asBoolean()
+                .orElse(true);
     }
 
+    // application class analysis
     @Override
     public AnalyzerResponse analyze(Class<?> maybeAnnotated) {
         AnalyzerResponse.Builder builder = AnalyzerResponse.builder();
 
         LoginConfig annotation = maybeAnnotated.getAnnotation(LoginConfig.class);
 
         if (null == annotation) {
+            builder.register(new RegisterMpJwt(false));
             return AnalyzerResponse.abstain();
         }
 
         if (isMpJwt(annotation)) {
-            return builder.authenticationResponse(Flag.REQUIRED)
+            // triggered!
+            builder.register(new RegisterMpJwt(true));
+
+            // now if there is a RolesAllowed on application class, automatically required, otherwise optional
+            // bugfix #455
+            // we may want to only authenticate requests that hit a @RolesAllowed annotated endpoint
+            Flag atnFlag = (secureByDefault ? Flag.REQUIRED : Flag.OPTIONAL);
+
+            if (isRolesAllowed(maybeAnnotated)) {
+                atnFlag = Flag.REQUIRED;
+            }
+
+            return builder.authenticationResponse(atnFlag)
                     .authenticator(authenticator)
                     .build();
         }
 
+        builder.register(new RegisterMpJwt(false));
         return builder.build();
     }
 
+    // resource class analysis
+    @Override
+    public AnalyzerResponse analyze(Class<?> maybeAnnotated, AnalyzerResponse previousResponse) {
+        if (isMpJwt(previousResponse) && isRolesAllowed(maybeAnnotated)) {
```

Comment:
Minor suggestion - Refactor this code and the code in the other variant of `analyze` into a separate private method to avoid duplicating it. I realize it's not a lot but it's about the same length as `isMpJwt` which is already in its own method.

Reply:
Refactored.